### ISSUE TITLE
Support moving across filesystems in pathlib.Path, as shutil.move() does

### ISSUE THREAD

**Issue by be1526b9-3573-4f3d-abe3-d392fe59c2a9** [python/cpython#73991]
BPO | [29805](https://bugs.python.org/issue29805)
--- | :---
Nosy | @brettcannon, @pfmoore, @ericvsmith, @tjguk, @zware, @eryksun, @zooba

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-03-13.21:03:44.694>
labels = ['3.8', 'type-feature', 'library', '3.9', '3.10']
title = 'Support moving across filesystems in pathlib.Path, as shutil.move() does'
updated_at = <Date 2021-03-15.21:26:30.922>
user = 'https://bugs.python.org/LaurentMazuel'
```

bugs.python.org fields:
```python
activity = <Date 2021-03-15.21:26:30.922>
actor = 'eryksun'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2017-03-13.21:03:44.694>
creator = 'Laurent.Mazuel'
dependencies = []
files = []
hgrepos = []
issue_num = 29805
keywords = []
message_count = 6.0
messages = ['289549', '289552', '289559', '289630', '289687', '289688']
nosy_count = 8.0
nosy_names = ['brett.cannon', 'paul.moore', 'eric.smith', 'tim.golden', 'Laurent.Mazuel', 'zach.ware', 'eryksun', 'steve.dower']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue29805'
versions = ['Python 3.8', 'Python 3.9', 'Python 3.10']
```

</p></details>

### Linked PRs
* gh-30650
* gh-119058
* gh-119060
* gh-119079
* gh-119159
* gh-120517
* gh-120519
* gh-120715
* gh-120718
* gh-120806
* gh-120807
* gh-121438
* gh-121444
* gh-122073
* gh-122368
* gh-122369
* gh-122924
* gh-123158
* gh-123314
* gh-123315
* gh-123337

**Comment by be1526b9-3573-4f3d-abe3-d392fe59c2a9** [comment]
Just to confirm, I was able to workaround it with Py3.6:

```py
    # client_generated_path.replace(destination_folder)
    shutil.move(client_generated_path, destination_folder)
```


It is reasonable to think about adding a similar feature to pathlib if it doesn't support it and just special-case the drive-to-drive scenario for Windows

**Comment by eryksun** [comment]
Moving a file across volumes isn't atomic. Getting an exception in this case can be useful. There could be a "strict" keyword-only parameter that defaults to False. If it's true, then replace() won't try to move the file.

**Comment by ericvsmith** [comment]
I agree this needs to be different from replace(), due to not being atomic. That makes it an enhancement, so I'm removing 3.5.

I'm +1 on Path supporting something like shutil.move().

**Comment by brettcannon** [comment]
I also support the idea of getting something like shutil.move() into pathlib.

**Comment by brettcannon** [comment]
I should also mention that rename() (https://docs.python.org/3/library/pathlib.html#pathlib.Path.rename) and replace() (https://docs.python.org/3/library/pathlib.html#pathlib.Path.replace) already do exist, so it might be best to add a keyword-only flag to one of those for this use-case.

**Comment by barneygale** [comment]
I'm also +1 on adding `pathlib.Path.move()`. **However** I'd prefer we *move* the `shutil` implementation into pathlib, [as proposed here](https://discuss.python.org/t/incrementally-move-high-level-path-operations-from-shutil-to-pathlib/19208). The implementation in `shutil` would become a stub that calls into pathlib, roughly:

```python
def move(src, dst, copy_function=copy2):
    return pathlib.Path(src).move(dst, copy_function=copy2)
```

I'm interested in this approach because I'm hoping to add a `pathlib.AbstractPath` class in future. Users would be able to subclass `AbstractPath`, and by implementing a small handful of low-level abstract methods like `stat()`, `open()` and `iterdir()`, benefit from high-level methods like `move()` in their objects. This would work even if the source and destination are different kinds of path (e.g. a local filesystem path and a path in a tar archive). 

To achieve this we'd need to make pathlib accept `bytes` paths, as shutil does. Personally I'm fine with this - it's not much work and perfectly reasonable.

We'd also need to move other shutil functions that `move()` relies on to pathlib, notably the `copy*()` functions, excluding `copyfileobj()` and friends.

On `rename()` / `replace()`: I'd like these to call through to `move(atomic=True)`. IIRC these methods are identical on POSIX and only differ slightly on Windows. That's a wrinkle we should smooth out in pathlib IMO.

**Comment by zooba** [comment]
> I'd prefer we _move_ the `shutil` implementation into pathlib

Sounds reasonable to me, though would we want to include `copy_function` in the API? It seems like an opportunity to say that `Path` is opinionated about the "most move-like way to copy", and if you need to override that then handle it yourself or stick with `shutil`.

**Comment by ericvsmith** [comment]
I think @barneygale 's plan is sound.

**Comment by pfmoore** [comment]
I agree with @zooba, exposing the `copy_function` in the `pathlib` API seems incompatible with the general style of `pathlib`. Otherwise the plan looks good.

**Comment by barneygale** [comment]
I agree about `copy_function`. Perhaps we could do something like this?

```python
    def copy(self, dst, *, follow_symlinks=True, copy_mode=True, copy_stat=False):
        ...

    def move(self, dst, *, atomic=False, **kwargs):
        if atomic:
            return os.rename(self, dst)
        self.copy(dst, **kwargs)
        self.unlink(self)
```

**Comment by zooba** [comment]
I don't like the `atomic` argument, and especially its default. Implementations should get to implement it as efficiently as possible without needing to respect that kind of argument, and callers shouldn't expect that level of control over semantics.

For example, on Windows we explicitly _disallow_ `os.rename` from moving files to other drives, because it wouldn't match our documented behaviour. But a new `WindowsPath.move` function should be able to allow that, and take advantage of the OS optimisations (yes, they exist!) for moving a file rather than copying it across drives.

If we were to define semantics for `atomic`, the only way to allow this would be "`atomic=True` will try to move, and `atomic=False` won't bother trying to move", at which point why would anybody ever pass False? Especially since there's no back-compat here, and so code on older versions will have to stick with `try: ...rename ... except: ... copy`.

We should make these functions an improvement in _semantics_, not just discoverability, and ideally more appropriate for whatever platform they're being used on than the POSIX semantics inherent to posixmodule.

So I think the only option we want on `move` is `follow_symlinks=True`,[^1] and the method "does whatever is necessary to move the file or directory as efficiently as possible into the new location while matching the semantics of an in-place rename" (roughly, `copy_mode=True, copy_stat=True`)

Similarly, I think we should simplify `copy` down to only have `follow_symlinks=True`,[^2] and the method "does whatever is necessary to copy the file or directory as efficiently as possible into the new location while matching the semantics of a newly created file" (roughly, `copy_mode=some bits, copy_stat=some bits` - and this one is why the vague definition is important 😉  )

[^1]: Where `follow_symlinks=False` is carefully defined to mean "if `self` is a link, it moves the link unless that would cause the new link to fail to refer to the same target as the original link, in which case the target is _copied_ into the new location and the original link is removed".

[^2]: I think the same definition as for move works here, but I haven't thought through this one as thoroughly.

**Comment by barneygale** [comment]
My example was a bit rubbish - here's a corrected version:

```python
    def copy(self, dst, *, follow_symlinks=True, copy_mode=True, copy_stat=False):
        ...

    def move(self, dst, *, atomic=False, **kwargs):
        try:
            return os.rename(self, dst)
        except OSError:
            if atomic:
                raise
        self.copy(dst, **kwargs)
        self.unlink()
```

So `move(atomic=True)` would call `os.rename()` only and allow exceptions to propagate, whereas `atomic=False` will try a rename, and failing that use a copy/delete. I think there's common use cases for both.

**Comment by barneygale** [comment]
I suppose one could argue that folks should just use `os.rename()` if they want an atomic rename! :)

**Comment by zooba** [comment]
> I suppose one could argue that folks should just use `os.rename()` if they want an atomic rename! :)

Yeah, I'll argue that one ;)

We'll have to explain that `move` may result in both old and new names existing simultaneously while the move takes place, depending on the operating system and file systems in use. That's a perfectly fine place to say that if you'd rather that _not_ happen, and are willing to accept exceptions if it's unavoidable, then you should use `rename`.

**Comment by barneygale** [comment]
Should we deprecate `pathlib.Path.rename()` and `replace()` when we add `move()`?

**Comment by barneygale** [comment]
That would be a change of behaviour though, right? `Path.rename()` currently raises an exception if you attempt to move across filesystems, whereas if we changed it to call `shutil.move()` it would not.

**Comment by barneygale** [comment]
Oh right. I thought we agreed that users could just call `os.rename()` or `os.replace()` if they specifically need that behaviour?

**Comment by zooba** [comment]
I think just add the new method, and don't deprecate any of the old ones. Moving people from `Path.rename()` to `os.rename()` feels like unnecessary churn to inflict.

We probably want to improve the docs of those to point people towards `move()` though, as we believe (don't we?) that it'll work more reliably in more circumstances. But there's no reason to force them.

**Comment by barneygale** [comment]
Thanks for the explanation!

Would you also rule out deprecating `Path.rename()` and `Path.replace()` in future? If so I'm -1 on adding `move()`. It's regrettable that we have two methods for moving files, and I wouldn't want to permanently add a third.

**Comment by zooba** [comment]
Technically, we have two functions for renaming files, and none for moving ;-)

I'd want to see it shown that they're a maintenance burden or an attractive nuisance[^1] before deprecating. Right now, I'm not convinced they are either (a bit of an attractive nuisance, but their long history outweighs that).

[^1]: Something that users assume will solve their issue, but it actually makes things worse for them.

**Comment by barneygale** [comment]
Re-visiting this issue!

Earlier I suggested that we should be able to move the `shutil.move()` implementation into `pathlib`. I now believe that won't work, because:

- pathlib's normalization can change the meaning of paths (i.e. strip trailing slashes)
- pathlib doesn't support `bytes`, whereas shutil does
- shutil's `copy_function` stuff doesn't fit well in pathlib

So I think that means we need to call `shutil.move()` from `pathlib.Path.move()`.

We now have a `pathlib._abc.PathBase` base class which ought to have a `move()` method too. That method can be implemented atop the existing `self.walk()`, `self.open()`, etc methods, though it means _also_ implementing versions of `copy()`, `copytree()` and `rmtree()`. These could be private to begin with, or we could add all four public methods at once.

We could make `pathlib.Path.move()` accept a `pathlib._abc.PathBase | os.PathLike` destination. If the destination is `os.PathLike` we call `shutil.move()`, otherwise we call `super().move()`. If we made `copy()` and `copytree()` public too, users would have a pretty versatile set of tools to copy/move files not just across local filesystems, but between entirely different types of filesystem. So:

```python
source_path = pathlib.Path('images')
target_path = s3.S3Path('web/images', bucket=..., client=...)
source_path.copytree(target_path)
```

**Comment by barneygale** [comment]
In `shutil.move()` and the `mv` shell command, if the destination is an existing directory, the source is copied *into* that directory.

Is this something we want to replicate in pathlib? It can be a source of bugs, e.g. if you `shutil.move('a', 'b')` and assume the new file is at 'b'.

Alternatively we could split it into `move()` and `move_into()` perhaps?

**Comment by zooba** [comment]
We shouldn't replicate that behaviour, but if the destination ends with a slash then I think we're okay to keep the original name and make that one work (pathlib paths won't ever end with a slash, I believe).

**Comment by barneygale** [comment]
PRs available:

- https://github.com/python/cpython/pull/119058
- https://github.com/python/cpython/pull/119060

Once those land I can add `copytree()` and finally `move()`

**Comment by serhiy-storchaka** [comment]
Is this a firm established? Because to me, adding the parts of `shutil` to `pathlib` looks very wrong. Why you cannot just use `shutil` directly? All this hassle with the path-like protocol was for making `pathlib` more usable with `os`, `shutil` and other modules without increasing coupling between modules. Adding the `shutil` dependency to `pathlib` will provide another argument to avoid using `pathlib`.

**Comment by zooba** [comment]
Without having looked at the details, I agree with Serhiy. We should avoid `import shutil` in modules that are "lower level" than it, which includes pathlib.

Passing a `Path` to shutil functions and having them work correctly (by calling back into pathlib, rather than assuming anything else about the underlying file system) is the important part. Maybe that already works fine? I haven't tried it with e.g. an S3 or Blob Storage implementation of Path.

But perhaps we should be adding copytree/rmtree methods to pathlib anyway that are self-contained? So the base implementation will work for anything that implements its `copy`, `glob` and `unlink` methods (I agree with `copy`, `move` and `rename` being added).

**Comment by barneygale** [comment]
> Is this a firm established? Because to me, adding the parts of `shutil` to `pathlib` looks very wrong. Why you cannot just use `shutil` directly? 

This is an argument against pathlib in general, and you're more than a decade too late unfortunately.

> All this hassle with the path-like protocol was for making `pathlib` more usable with `os`, `shutil` and other modules without increasing coupling between modules.

The `os.PathLike` stuff meant that `os` didn't need to import `pathlib`, but you seem to be suggesting the opposite? This patch doesn't make `shutil` import `pathlib`.

> Adding the `shutil` dependency to `pathlib` will provide another argument to avoid using `pathlib`.

Explain that argument, please? Increased import time or something?

**Comment by barneygale** [comment]
> Passing a `Path` to shutil functions and having them work correctly (by calling back into pathlib, rather than assuming anything else about the underlying file system) is the important part. Maybe that already works fine? I haven't tried it with e.g. an S3 or Blob Storage implementation of Path.

Pathlib strips important information (e.g. leading `./`, trailing `/`) so we can't implement `shutil.rmtree()` using `Path.rmtree()`, because we'd get different behaviour. More here: https://discuss.python.org/t/pathlib-preserve-trailing-slash/33389

We could maybe add a `RawUnnormalizingLocalPath` class that doesn't perform normalization, which we could employ from `shutil.rmtree()`.

**Comment by barneygale** [comment]
There's some more discussion here: https://discuss.python.org/t/incrementally-move-high-level-path-operations-from-shutil-to-pathlib/19208

I might revive that thread as it covers exactly this ticket. Perhaps adding these methods is only compelling as part of making `PathBase` public.

**Comment by zooba** [comment]
> so we can't implement `shutil.rmtree()` using `Path.rmtree()`, because we'd get different behaviour.

Should we implement them separately, then? I suspect `shutil` is only going to work for "normal" filesystem paths, and so the best thing to do is for `pathlib` to have its own implementations in terms of its own functions, and for `shutil` to use `__fspath__` and reject anything that doesn't return a file system path?

**Comment by barneygale** [comment]
> Should we implement them separately, then? 

That's what I'm intending to do in the `PathBase` implementations - they won't use `shutil` except for `copyfileobj()`. All the implementations will be generic and built on lower-level methods like `open()`, `symlink_to()`, etc.

The story is more complicated in `Path`:

- `copy()` should call `shutil.copy2()` if its argument is `os.PathLike`, because `shutil.copy2()` has extensive low-level code to properly copy metadata
- `rmtree()` should probably call `shutil.rmtree()` to make use of its fd-based safe traversal where available. Alternatively I could add `dir_fd` support to `Path.walk()`, but it's a bit of a can of worms
- `copytree()` can be inherited from `PathBase` (it uses `walk()`, `mkdir()` and `copy()`)
- `move()` can also be inherited from `PathBase` (it uses `copytree()`, `rmtree()`, etc)

**Comment by zooba** [comment]
> * `copy()` should call `shutil.copy2()` if its argument is `os.PathLike`, because `shutil.copy2()` has extensive low-level code to properly copy metadata

It's not extensive, there's just a choice between using `open` and `_winapi.CopyFile` on Windows and some error handling. Having separate implementations on `WindowsPath` and `PosixPath` ought to make this look much simpler (though of course, we need to fall back on a read/write even in the Windows case, for implementations that may not replicate all of `_winapi`).

> * `rmtree()` should probably call `shutil.rmtree()` to make use of its fd-based safe traversal where available.

It can do this for `WindowsPath` and `PosixPath`, but the base implementation shouldn't use `shutil` because it can't be sure it's dealing with a file system. Using `walk`, `unlink` and `rmdir` here ought to be sufficient, and retrying a failed `rmdir` by re-walking the directory is probably as robust as it needs to be.

> * `move()` can also be inherited from `PathBase` (it uses `copytree()`, `rmtree()`, etc)

And we should also (not necessarily in this issue) add a `WindowsPath` override that uses [`MoveFileEx`](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-movefileexw), which can already do cross-volume moves if requested.

**Comment by barneygale** [comment]
> > * `copy()` should call `shutil.copy2()` if its argument is `os.PathLike`, because `shutil.copy2()` has extensive low-level code to properly copy metadata
> 
> It's not extensive, there's just a choice between using `open` and `_winapi.CopyFile` on Windows and some error handling.

It can call through to `copystat()`, which does more. It feels pretty icky to copy-paste all that into pathlib, right? Maybe I'm overreacting.

> > * `rmtree()` should probably call `shutil.rmtree()` to make use of its fd-based safe traversal where available.
> 
> It can do this for `WindowsPath` and `PosixPath`, but the base implementation shouldn't use `shutil` because it can't be sure it's dealing with a file system. Using `walk`, `unlink` and `rmdir` here ought to be sufficient, and retrying a failed `rmdir` by re-walking the directory is probably as robust as it needs to be.

Yes that's what I mean, sorry for not being clear. `Path.rmtree()` calls `shutil.rmtree()`, but `PathBase.rmtree()` is a generic implementation atop `walk()`, `unlink()` and `rmdir()`. This is implemented in #119060

**Comment by zooba** [comment]
> It can call through to `copystat()`, which does more.

Yeah, `copystat` is a bit nasty. Maybe it's something that deserves a (private?) `os` implementation? It seems more in line with that module to me.

Most of these functions seem to be careful error handling. I'm not sure we can do anything _but_ copy paste all that.

**Comment by barneygale** [comment]
> Most of these functions seem to be careful error handling. I'm not sure we can do anything _but_ copy paste all that.

Couldn't we call `shutil.copy2()` from `Path.copy()` if the target is `os.PathLike`? That's how I've implemented it in #119058, and it avoids any copy-pasting!

**Comment by zooba** [comment]
`os.PathLike` is the criteria used by things outside of pathlib, not by pathlib itself, so I hope you're just using it casually. pathlib ought to know whether its own subclasses are for file systems or not, and if it's a user subclass, then the implementer should override.

But the main objection is to the architectural coupling between the modules - `shutil` should not be a dependency of `pathlib`, just as it should not be a dependency of `os.path`. The direction of the relationship flows the other way.

The best way to deal with it would be to move the implementations from `shutil` into `pathlib` and then update `shutil` to call them there (or move them into `os` or `os.path`).

**Comment by barneygale** [comment]
> But the main objection is to the architectural coupling between the modules - `shutil` should not be a dependency of `pathlib`, just as it should not be a dependency of `os.path`. The direction of the relationship flows the other way.

I agree that `shutil` is higher-level than `os`, but I don't agree that `shutil` is also higher level than `pathlib`. I think they're mostly on about the same conceptual level, providing "high level" functions built on lower-level building blocks. e.g. pathlib has `Path.glob()` and `Path.walk()` that do recursive directory walking using low-level `scandir()`, and shutil has `copytree()` and `rmtree()` that do the same.

In some cases `shutil` is higher-level, particularly its archiving operations.

In other cases it's slightly lower-level IMO, particularly the implementations of `copyfile()`, `copystat()` and friends, which wouldn't seem out-of-place in `os`, and deal with low-level facets of the local filesystem (BSD symlink permissions, xattrs etc) that pathlib doesn't currently touch.

Well, that's my current view anyway. You have decades more experience on this than I do, so I what it looks like to implement this without pathlib importing shutil :)

**Comment by pfmoore** [comment]
Very off-topic, but I seem to only be getting email notifications of @barneygale's comments, and nothing for any of @zooba's. Which makes following the conversation weirdly difficult... Does anyone know how this could even happen? I have checked, and Steve's messages aren't ending up in my spam folder, so I don't think it's a problem with my email client.

**Comment by zooba** [comment]
> I seem to only be getting email notifications of @barneygale's comments, and nothing for any of @zooba's.

I wouldn't blame you for filtering my messages 😆 But no, I'm afraid I don't know of any reason other than that.

> Just to be clear, do you mean that `shutil` would call new _module functions_ in `pathlib`, or that `shutil` would instantiate some kind of `Path` object and call appropriate methods?

I was deliberately non-specific, because I don't know off the top of my head which would be best and I trust you to figure it out.

I'd half hoped that instantiating `Path` from shutil would be enough, but if you say it's not, then I'm a bit sad but I believe you. Bytes paths should be round-tripping through `surrogateescape` though, so there shouldn't be any loss. I do recall the discussions about trailing backslashes though.

Now you've shown what would have to happen to the hierarchy, I agree that module-level helper functions are probably best, and putting them in `os` probably makes even more sense. That module has already gone so far beyond exposing merely what POSIX specifies that I don't think we're breaking anything architecturally there, and the functions ought to only rely on `os` and `os.path` functions anyway.

**Comment by barneygale** [comment]
Trailing slashes are indeed the sticking point :-(. If I could go back in time I'd ask Antoine to retain them somehow when he designed pathlib, then go kill baby hitler etc.

I like the idea of moving the lowest-level stuff into `os` or `os.path`. I'll try that :-)

**Comment by pfmoore** [comment]
Weird. I got the notification for that. So the problem is just on this issue (or maybe this project). I'll do some investigating.

**Comment by barneygale** [comment]
Continued on the forum: https://discuss.python.org/t/incrementally-move-high-level-path-operations-from-shutil-to-pathlib/19208/32

**Comment by barneygale** [comment]
State of my todo list:

- [x] Add `Path.copy()`
- [ ] Add support for not following symlinks when copying
- [ ] Add support for preserving specific metadata when copying
- [ ] Add `Path.copytree()`
- [ ] Add `Path.rmtree()`
- [ ] Add `Path.move()`

**Comment by barneygale** [comment]
To implement `Path.move()` I think we'll need to support copying available metadata via argument(s) to `Path.copy()`.

On my Linux system, `cp --preserve` supports 'mode', 'ownership', 'timestamps' (the defaults), and 'context', 'links', 'xattr', 'all'. Translated into pathlib I guess we could add an enum or some constants, so:

```python
source.copy(target, flags=pathlib.COPY_MODE | pathlib.COPY_XATTR | etc)
```

Or we could keep it simple:

```python
source.copy(target, metadata=True)
```

Where `metadata=True` means "copy whatever metadata you can please".

Or something else. Thoughts?

**Comment by zooba** [comment]
> To implement `Path.move()` I think we'll need to support copying available metadata via argument(s) to `Path.copy()`.

Why? Because moving within a filesystem preserves it all but copying doesn't?

On designs, I prefer `metadata=True` in this case, because I'd rather a big "copy mode" option so we can still do-what-I-mean.

I'd also suggest understanding/deciding how it will apply to other filesystems and/or targets that pathlib may support, so we can at least document the option in terms that _might_ apply to those as well. "Like `cp --preserve`" is not a good definition.

**Comment by barneygale** [comment]
Hey @serhiy-storchaka, per your [earlier comment](https://github.com/python/cpython/pull/119079#issuecomment-2114341003) I've been careful not to use `shutil` from `pathlib` so far. If it's OK with you, I'd like to move some private `shutil` functionality into a `pathlib._os` module such that it can be used by both modules. Specifically:

- GH-119060 moves the `rmtree()` implementations (safe + unsafe)
- GH-120806 moves the `copystat()` implementation

Does this sound OK to you and address your concerns? If you're not a fan of the `pathlib._os` module path we could switch to `_pathlib_os` or `_shutil_impl` or something like that? Thanks!

**Comment by barneygale** [comment]
Alternatively, we could keep those implementations in `shutil.py` and use them from pathlib, e.g.:

```python
# Lib/pathlib/_local.py

from shutil import _rmtree_impl
```

... which is a simpler patch that preserve the `git blame`, but feels a bit gross to me - better a shared private module than dipping into private functionality of a public module?

**Comment by encukou** [comment]
Since GH-120718, the “AMD64 Debian root 3.x” buildbot (which runs tests under the root account) has been failing with `AssertionError: PermissionError not raised by copytree`.

**Comment by zooba** [comment]
> ```python
> Lib/pathlib/_local.py
> 
> from shutil import
> ```

This bit feels gross to me. But equally, so does `shutil.py // from pathlib ...`. I'd really be inclined to just have a copy if we're not able to agree on putting private methods in a low level module.

That said, importing `pathlib` from `shutil` feels _less_ bad than going the other way. So I'd vote for that direction if we have to do either.

**Comment by serhiy-storchaka** [comment]
I think that `pathlib` and `shutil` should be independent modules. I agree that importing `pathlib` from `shutil` is less bad that other way, but still...

If you insist on adding more and more stuff in `pathlib`, I suggest to consider an option of splitting `pathlib` on `simplepathlib`, which only supports path manipulation operations and simple OS operations, and `richpathlib`, which can import `shutil`, `tarfile`, `json`, whatever.

**Comment by barneygale** [comment]
Obviously I am not going to split pathlib into `simplepathlib` and `richpathlib`, and I'm sure you didn't mean it as a serious suggestion.

**Comment by zooba** [comment]
Even if it were split, I'd argue that recursively removing a directory tree and moving across filesystems is `simplepathlib` anyway, so it doesn't help here. Putting `rmtree` in `shutil` was the original mistake, and I'm sure it made perfect sense at the time (there's no C API for it).

These days, the concept of "simple OS operation" is defined by the shell, not by libc. Modern users come in and expect shell-like functionality to be a natural part of our libraries, rather than something special to opt into. Possibly those who come in from a C only background would still think of a separate library, but that's not really the pathlib audience.

**Comment by barneygale** [comment]
> I'd really be inclined to just have a copy if we're not able to agree on putting private methods in a low level module.

Copying the `shutil.copystat()` code is fairly reasonable - in my PR I've split it into `pathlib._os.get_file_metadata()` and `set_file_metadata()` methods, which is a bit slower for shutil users.

It's harder to find a good justification to copy the `shutil.rmtree()` code into pathlib, as it already does exactly what we want - no more, no less - and its implementation is rather technical and complex in places. Repeating that code verbatim in pathlib is gross in a different way IMO.

FWIW, we should be able to import and use `shutil.rmtree()` directly from pathlib if we wanted. That's certainly preferable to pathlib importing *private* functionality from shutil, but probably still inferior to a single low-level module from which both shutil and pathlib import?

**Comment by zooba** [comment]
> That's certainly preferable to pathlib importing _private_ functionality from shutil, but probably still inferior to a single low-level module from which both shutil and pathlib import?

I'd be okay with a dynamic import (i.e. import it in the remove function, not at module load time). I suspect it's only a matter of time until we end up having to decide whether certain fixes need to affect both or just one path, but at least if it's the same function we've got a strong argument for needing fixes in `shutil`.

And subclasses of `Path` still get to override it for their own platform, so they're happy (and I'm happy).

**Comment by barneygale** [comment]
Cool, I like that. I've adjusted #120806 and #119060 along those lines.

The only downside is that we lack a `pathlib.Path.rmtree.avoids_symlink_attacks` attribute analogous to [`shutil.rmtree.avoids_symlink_attacks`](https://docs.python.org/3/library/shutil.html#shutil.rmtree.avoids_symlink_attacks), but it's not vital to expose right now.

**Comment by barneygale** [comment]
Q: should `Path.move()` support both files and directories, or should we support files in `move()` and directories in `movetree()`? So far I've maintained separate methods for dirs and non-dirs (`copy()` and `copytree()`, `unlink()` and `rmtree()`), but `shutil.move()` handles both.

Given we're not necessarily aiming for compatibility with the `mv` shell command, I reckon separating them might be best?

Or we could go in the opposite direction: support both files and directories in `move()`, `copy()` and a new `remove()` function.

**Comment by barneygale** [comment]
I've been closely following the `shutil` API so far, resulting in this:

|                 | **delete** | **copy**     | **move** |
|-----------------|------------|--------------|----------|
| **file**        | unlink()   | _copy()_     |          |
| **dir (empty)** | rmdir()    |              |          |
| **dir (any)**   | _rmtree()_ | _copytree()_ |          |
| **anything**    |            |              | _move()_ |

(new methods shown in italics)

Having given it some thought, I have the following concerns:

1. `copy()` and `move()` are the shortest/most obvious methods, but the former accepts only files, whereas the latter accepts both files and directories
2. `move()` can be used to move any kind of file, but there are no corresponding methods for copying and deletion.

And so I'd like to propose that:

1. We rename `copy()` to `copyfile()`
2. We rename `rmtree()` to `delete()`, and `copytree()` to `copy()`, and add support for files to both methods.

This would result in:

|                 | **delete** | **copy**     | **move** |
|-----------------|------------|--------------|----------|
| **file**        | unlink()   | _copyfile()_ |          |
| **dir (empty)** | rmdir()    |              |          |
| **dir (any)**   |            |              |          |
| **anything**    | _delete()_ | _copy()_     | _move()_ |

Thoughts?

**Comment by barneygale** [comment]
(Also, if we wanted to add support for moving/copying *into an existing directory* via new methods, it would be better to have `move_into()` and `copy_into()`, rather than `move_into()`, `copyfile_into()` and `copytree_into()` IMO.)

**Comment by pfmoore** [comment]
+1 from me on the revised names and the `_into` forms.

Do we even need a `copyfile()` function, if `copy()` will do the same?

**Comment by barneygale** [comment]
> Do we even need a `copyfile()` function, if `copy()` will do the same?

Good point - perhaps not!

**Comment by barneygale** [comment]
It's done! Thank you everyone who helped formulate and review this feature.

We've added new `copy()`, `copy_into()`, `move()` and `move_into()` methods, which work on both files and directories. The interfaces are refined to pathlib standards, with certain nonessential functionality (like support for filtering in `copy()`) omitted for now. Users may wish to propose new capabilities on the [ideas forum](https://discuss.python.org/c/ideas/6).

The new methods expect exact targets (or target directories for the `_into()` variants), whereas `shutil` checks for existing directories and adjusts the target as necessary. The latter behaviour is helpful in an interactive shell with tab-complete, but less so in a stored program where state of the filesystem might not be known (for example, `shutil.copy('a', 'b')` could create either `b` or `b/a`).

The `copy()` and `copy_into()` methods use `fcntl.FICLONE` or `os.copy_file_range` where available (see [GH-81338](https://github.com/python/cpython/issues/81338), [GH-81340](https://github.com/python/cpython/issues/81340)). This work has stalled for years in `shutil` for backwards-compatibility concerns, happily not present here.

Feedback most welcome! We have ~9 months until 3.14 beta 1 where the interface is more locked down, so there's still plenty of time make changes. I'll close this issue, but I can re-open it (or log new issues) if problems are identified. Thanks again, all.

**Comment by encukou** [comment]
For copying trees, what happens on error -- for example, when one of the files has a name that's incompatible with the target filesystem?

**Comment by barneygale** [comment]
In that case you'll get an `OSError`. Any copies already made will be left on disk, and in the case of `move()`, none of the original files will be deleted.

**Comment by agronholm** [comment]
It looks like, unlike `copy()`, this method fails if I pass a path-like object that just *happens* to have a method named `stat()`. It then fails at `self._ensure_different_file(target)` which already assumes a `Path` as its argument type. It really should do an isinstance check on the target. In my case, `stat()` is a coroutine method, so it fails with `AttributeError: 'coroutine' object has no attribute 'st_ino'`.

**Comment by zooba** [comment]
Passing an object that looks similar but isn't the same is the caller's error. An `AttributeError` is exactly what you should get in this case.

That said, it probably wouldn't be unreasonable to convert any incoming paths to the same type as the source. We shouldn't type check here, we should try to convert and let the errors bubble out to the caller.

**Comment by agronholm** [comment]
Such type checks are done everywhere else, **including** in the method in question! This would be solved simply by doing said type check earlier.

Is it unreasonable to expect this to fail on a `PathLike` object? How am I supposed to know what I can pass as the target then?

**Comment by zooba** [comment]
In general for pathlib, you should be able to pass any object that can be interpreted as the same kind of path as the one you're passing it to. I know that upsets the static typing people because it's hard to define, but being accepting of inputs is a key Python feature.

In this case, that means converting the input to the same type as the target before using it. Most of the functions that have type checks are essentially doing this - they're _negative_ checks, followed by essentially `type(self)(other)`, which will try its hardest to parse the path out of the input. But `samefile` is not doing this and clearly ought to be. I suggest opening a new issue.
